src/libbson/src/bson/bson-context.c

```diff
@@ -264,79 +142,236 @@ _bson_context_get_hostname (char *out)
 }
 
 
-static void
-_bson_context_init_random (bson_context_t *context, bool init_sequence)
+#define ROTL(x, b) (uint64_t) (((x) << (b)) | ((x) >> (64 - (b))))
+
+#define U32TO8_LE(p, v)            \
+   (p)[0] = (uint8_t) ((v));       \
+   (p)[1] = (uint8_t) ((v) >> 8);  \
+   (p)[2] = (uint8_t) ((v) >> 16); \
+   (p)[3] = (uint8_t) ((v) >> 24);
+
+#define U64TO8_LE(p, v)               \
+   U32TO8_LE ((p), (uint32_t) ((v))); \
+   U32TO8_LE ((p) + 4, (uint32_t) ((v) >> 32));
+
+#define U8TO64_LE(p)                                            \
+   (((uint64_t) ((p)[0])) | ((uint64_t) ((p)[1]) << 8) |        \
+    ((uint64_t) ((p)[2]) << 16) | ((uint64_t) ((p)[3]) << 24) | \
+    ((uint64_t) ((p)[4]) << 32) | ((uint64_t) ((p)[5]) << 40) | \
+    ((uint64_t) ((p)[6]) << 48) | ((uint64_t) ((p)[7]) << 56))
+
+void
+_sip_round (uint64_t *v0, uint64_t *v1, uint64_t *v2, uint64_t *v3)
 {
-   int64_t rand_bytes;
-   struct timeval tv;
-   unsigned int seed = 0;
-   char hostname[HOST_NAME_MAX];
-   char *ptr;
-   int hostname_chars_left;
-
-   /*
-    * The seed consists of the following xor'd together:
-    * - current time in seconds
-    * - current time in milliseconds
-    * - current pid
-    * - current hostname
-    */
-   bson_gettimeofday (&tv);
-   seed ^= (unsigned int) tv.tv_sec;
-   seed ^= (unsigned int) tv.tv_usec;
-   seed ^= (unsigned int) context->pid;
-
-   context->gethostname (hostname);
-   hostname_chars_left = strlen (hostname);
-   ptr = hostname;
-   while (hostname_chars_left) {
-      uint32_t hostname_chunk = 0;
-      uint32_t to_copy = hostname_chars_left > 4 ? 4 : hostname_chars_left;
-
-      memcpy (&hostname_chunk, ptr, to_copy);
-      seed ^= (unsigned int) hostname_chunk;
-      hostname_chars_left -= to_copy;
-      ptr += to_copy;
-   }
+   *v0 += *v1;
+   *v1 = ROTL (*v1, 13);
+   *v1 ^= *v0;
+   *v0 = ROTL (*v0, 32);
+   *v2 += *v3;
+   *v3 = ROTL (*v3, 16);
+   *v3 ^= *v2;
+   *v0 += *v3;
+   *v3 = ROTL (*v3, 21);
+   *v3 ^= *v0;
+   *v2 += *v1;
+   *v1 = ROTL (*v1, 17);
+   *v1 ^= *v2;
+   *v2 = ROTL (*v2, 32);
+}
 
-#ifndef BSON_HAVE_RAND_R
-   srand (seed);
-#endif
+void
+_siphash (const void *in,
+          const size_t inlen,
+          const uint64_t key[2],
+          uint64_t digest[2])
```

Comment:
Was a reference used to implement this function (e.g. https://eprint.iacr.org/2012/351.pdf)? If so, please add a source as comment/documentation.

Reply:
Updated with a reference